### PR TITLE
Add lily58 datadavd layout

### PR DESCRIPTION
**Pull request by DataDavD**
# Adds my 4 layer custom layout to official QMK repo under my GitHub username

## Types of Changes
- Add 4 layer keymap for Lily58 Pro
- Customize OLED images

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* N/A

## Checklist
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/lily58/lib/layer_state_reader.c [hunk 1]
```diff
@@ -4,10 +4,10 @@
 #include "lily58.h"
 
 #define L_BASE 0
-#define L_LOWER (1 << 1)
+#define L_SUPER (1 << 1)
```
Comment (drashna) [inline keyboards/lily58/lib/layer_state_reader.c:7]: please revert these changes. 

If you need the changes, then the contents should be copied to your keymap, instead.

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by DataDavD** [comment]
Thank you  @drashna!!!! ill get on these this week.

File: keyboards/lily58/lib/layer_state_reader.c (hunk 1 above)
Comment (DataDavD) [reply keyboards/lily58/lib/layer_state_reader.c:7]: hey @drashna I made the suggested changes except for this one. Sorry to ask, but I am curious how/why the keycap specific layer_state_reader should be copied to my keymap. Will a `layer_state_reader.c` file in my keymap dir overwrite the keyboard/lily58 layer_state_reader???? Thanks in advance.

Comment (drashna) [reply keyboards/lily58/lib/layer_state_reader.c:7]: No, it wouldn't.  

And honestly, I'm not a fan of the `lib` files, in general.  It's not flexible, and would be better suited for the keymap file, in general.

Comment (DataDavD) [reply keyboards/lily58/lib/layer_state_reader.c:7]: Ok I see. So, I should

1. revert this layer_state_reader
2. move my code from this layer_state_reader to my keycap file

Sound correct?

Thanks again for your guidance. I really appreciate it.

Best,
David

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/lily58/keymaps/datadavd/keymap.c [hunk 2]
```diff
@@ -0,0 +1,247 @@
+/* Copyright %YEAR% %YOUR_NAME%
```
[the remark this answers is not in the document]
Comment (DataDavD) [reply keyboards/lily58/keymaps/datadavd/keymap.c]: @drashna done.

File: keyboards/lily58/lib/layer_state_reader.c (hunk 1 above)
Comment (DataDavD) [reply keyboards/lily58/lib/layer_state_reader.c]: @drashna I reverted the layer state reader and put the keycap specific layer state reader code within my keymap. Is that right?

Comment (drashna) [reply keyboards/lily58/lib/layer_state_reader.c]: Yup, thanks!

File: keyboards/lily58/keymaps/datadavd/rules.mk [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+SLEEP_LED_ENABLE = no    # Breathing sleep LED during USB suspend
+
+# If you want to change the display of OLED, you need to change here
+SRC +=  ./lib/rgb_state_reader.c \
+        ./lib/layer_state_reader.c \
+        ./lib/logo_reader.c \
+        # ./lib/keylogger.c \
+        # ./lib/mode_icon_reader.c \
+        # ./lib/host_led_state_reader.c \
+        # ./lib/timelogger.c \
```
Comment (fauxpark) [inline keyboards/lily58/keymaps/datadavd/rules.mk:30]:
```suggestion
MOUSEKEY_ENABLE = yes
EXTRAKEY_ENABLE = yes

# If you want to change the display of OLED, you need to change here
SRC +=  ./lib/rgb_state_reader.c \
        ./lib/layer_state_reader.c \
        ./lib/logo_reader.c
```

File: keyboards/lily58/keymaps/datadavd/keymap.c [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+ *                   |      |      |      |/       /         \      \ |      |      |      |
+ *                   `----------------------------'           '------''--------------------'
+ */
+
+ [_QWERTY] = LAYOUT( \
+  KC_ESC,   KC_1,   KC_2,    KC_3,    KC_4,    KC_5,                     KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_BSPC, \
+  KC_TAB,   KC_Q,   KC_W,    KC_E,    KC_R,    KC_T,                     KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_BSLS, \
+  KC_LSFT,  KC_A,   KC_S,    KC_D,    KC_F,    KC_G,                     KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT, \
+  KC_LCTRL, KC_Z,   KC_X,    KC_C,    KC_V,    KC_B,   KC_GRV, KC_ENT,   KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH,  KC_RSFT, \
+                        KC_LBRC, KC_LALT, KC_LGUI, MO(_SUPER), KC_SPC, MO(_RAISE), KC_RGUI, KC_RBRC \
```
Comment (fauxpark) [inline keyboards/lily58/keymaps/datadavd/keymap.c:48]:
```suggestion
 [_QWERTY] = LAYOUT(
  KC_ESC,   KC_1,   KC_2,    KC_3,    KC_4,    KC_5,                     KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_BSPC,
  KC_TAB,   KC_Q,   KC_W,    KC_E,    KC_R,    KC_T,                     KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_BSLS,
  KC_LSFT,  KC_A,   KC_S,    KC_D,    KC_F,    KC_G,                     KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT,
  KC_LCTRL, KC_Z,   KC_X,    KC_C,    KC_V,    KC_B,   KC_GRV, KC_ENT,   KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH,  KC_RSFT,
                        KC_LBRC, KC_LALT, KC_LGUI, MO(_SUPER), KC_SPC, MO(_RAISE), KC_RGUI, KC_RBRC
```

File: keyboards/lily58/keymaps/datadavd/keymap.c [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+ *                   |Trans | LAlt |LGUI  | /Trans  /       \Space \  |Del   | TGUI |Trans |
+ *                   |      |      |      |/       /         \      \ |      |      |      |
+ *                   `----------------------------'           '------''--------------------'
+ */
+[_SUPER] = LAYOUT( \
+  KC_POWER,  KC_PWR,   KC_SLEP,  KC_WAKE,  KC_NO,     KC_NO,                         KC_BSLS,  KC_GRV,   KC_RCBR,  KC_MINS,  KC_EQL,  KC_BSPC,\
+  KC_TAB,    KC_TAB,   KC_ESC,   KC_ENT,   KC_LCTRL,  KC_LCBR,                       KC_RPRN,  KC_TAB,   KC_UP,    KC_LBRC,  KC_RBRC, KC_TRNS, \
+  KC_LSFT,   KC_LSFT,  KC_LGUI,  KC_LALT,  KC_UNDS,   KC_LPRN,                       KC_BSPC,  KC_LEFT,  KC_DOWN,  KC_RGHT,  KC_EQL,  KC_ENT, \
+  KC_LCTL,   KC_LCTL,  KC_HOME,  KC_PGUP,  KC_PGDN,   KC_END,   KC_CAPS,  KC_GRV,    KC_B,     KC_V,     KC_C,     KC_X,     KC_Z,    KC_RSFT, \
+                                   KC_TRNS, KC_LALT, KC_LGUI, KC_TRNS,      KC_SPC, KC_DEL, KC_RGUI, KC_TRNS\
```
Comment (fauxpark) [inline keyboards/lily58/keymaps/datadavd/keymap.c:69]:
```suggestion
[_SUPER] = LAYOUT(
  KC_POWER,  KC_PWR,   KC_SLEP,  KC_WAKE,  KC_NO,     KC_NO,                         KC_BSLS,  KC_GRV,   KC_RCBR,  KC_MINS,  KC_EQL,  KC_BSPC,
  KC_TAB,    KC_TAB,   KC_ESC,   KC_ENT,   KC_LCTRL,  KC_LCBR,                       KC_RPRN,  KC_TAB,   KC_UP,    KC_LBRC,  KC_RBRC, KC_TRNS,
  KC_LSFT,   KC_LSFT,  KC_LGUI,  KC_LALT,  KC_UNDS,   KC_LPRN,                       KC_BSPC,  KC_LEFT,  KC_DOWN,  KC_RGHT,  KC_EQL,  KC_ENT,
  KC_LCTL,   KC_LCTL,  KC_HOME,  KC_PGUP,  KC_PGDN,   KC_END,   KC_CAPS,  KC_GRV,    KC_B,     KC_V,     KC_C,     KC_X,     KC_Z,    KC_RSFT,
                                   KC_TRNS, KC_LALT, KC_LGUI, KC_TRNS,      KC_SPC, KC_DEL, KC_RGUI, KC_TRNS
```

File: keyboards/lily58/keymaps/datadavd/keymap.c [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+ *                   |      |      |      |/       /         \      \ |      |      |      |
+ *                   `----------------------------'           '------''--------------------'
+ */
+
+[_RAISE] = LAYOUT( \
+  KC_F1,   KC_F2,   KC_F3,   KC_F4,    KC_F5,    KC_F6,                         KC_F7,    KC_F8,    KC_F9,   KC_F10,      KC_F11,      KC_F12, \
+  KC_TAB,  KC_TAB,  KC_NO,   KC_LPRN,  KC_RPRN,  KC_TAB,                        KC_MUTE,  KC_VOLU,  KC_UP,   MO(_MOUSE),  KC_ENT,      KC_CAPS, \
+  KC_LSFT, KC_LSFT, KC_DEL,  KC_LBRC,  KC_RBRC,  KC_GRV,                        KC_VOLD,  KC_LEFT,  KC_DOWN, KC_RGHT,     KC_RSFT,     MO(_MOUSE), \
+  KC_LCTL, KC_LCTL, KC_MRWD, KC_MPLY,  KC_MFFD,  KC_MSTP,   KC_BRID, KC_BRIU,   KC_END,   KC_PGDN,  KC_PGUP, KC_HOME,     KC_RSFT,     KC_RSFT, \
+                               KC_NO, KC_LALT, KC_LGUI, KC_BSPC,      KC_TRNS, KC_TRNS, KC_TRNS, KC_NO \
```
Comment (fauxpark) [inline keyboards/lily58/keymaps/datadavd/keymap.c:91]:
```suggestion
[_RAISE] = LAYOUT(
  KC_F1,   KC_F2,   KC_F3,   KC_F4,    KC_F5,    KC_F6,                         KC_F7,    KC_F8,    KC_F9,   KC_F10,      KC_F11,      KC_F12,
  KC_TAB,  KC_TAB,  KC_NO,   KC_LPRN,  KC_RPRN,  KC_TAB,                        KC_MUTE,  KC_VOLU,  KC_UP,   MO(_MOUSE),  KC_ENT,      KC_CAPS,
  KC_LSFT, KC_LSFT, KC_DEL,  KC_LBRC,  KC_RBRC,  KC_GRV,                        KC_VOLD,  KC_LEFT,  KC_DOWN, KC_RGHT,     KC_RSFT,     MO(_MOUSE),
  KC_LCTL, KC_LCTL, KC_MRWD, KC_MPLY,  KC_MFFD,  KC_MSTP,   KC_BRID, KC_BRIU,   KC_END,   KC_PGDN,  KC_PGUP, KC_HOME,     KC_RSFT,     KC_RSFT,
                               KC_NO, KC_LALT, KC_LGUI, KC_BSPC,      KC_TRNS, KC_TRNS, KC_TRNS, KC_NO
```

File: keyboards/lily58/keymaps/datadavd/keymap.c [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+ *                   |      |      |      |/       /         \      \ |      |      |      |
+ *                   `----------------------------'           '------''--------------------'
+ */
+
+[_MOUSE] = LAYOUT( \
+  KC_NO,   KC_NO,   KC_NO,   KC_NO,    KC_NO,    KC_NO,                     KC_NO,  KC_NO,  KC_NO,    KC_NO,    KC_NO,    KC_NO, \
+  KC_NO,   KC_NO,   KC_WH_D, KC_MS_U,  KC_WH_U,  KC_NO,                     KC_NO,  KC_NO,  KC_LSFT,  KC_TRNS,  KC_LGUI,  KC_NO, \
+  KC_NO,   KC_BTN2, KC_MS_L, KC_MS_D,  KC_MS_R,  KC_NO,                     KC_NO,  KC_NO,  KC_NO,    KC_NO,    KC_NO,    KC_TRNS, \
+  KC_NO,   KC_NO,   KC_WH_R, KC_BTN3,  KC_WH_L,  KC_NO,    KC_NO, KC_NO,    KC_NO,  KC_NO,  KC_NO,    KC_NO,    KC_NO,    KC_NO, \
+                                  KC_NO, KC_NO, KC_NO, KC_BTN1,    KC_NO, KC_TRNS, KC_NO, KC_NO \
```
Comment (fauxpark) [inline keyboards/lily58/keymaps/datadavd/keymap.c:113]:
```suggestion
[_MOUSE] = LAYOUT(
  KC_NO,   KC_NO,   KC_NO,   KC_NO,    KC_NO,    KC_NO,                     KC_NO,  KC_NO,  KC_NO,    KC_NO,    KC_NO,    KC_NO,
  KC_NO,   KC_NO,   KC_WH_D, KC_MS_U,  KC_WH_U,  KC_NO,                     KC_NO,  KC_NO,  KC_LSFT,  KC_TRNS,  KC_LGUI,  KC_NO,
  KC_NO,   KC_BTN2, KC_MS_L, KC_MS_D,  KC_MS_R,  KC_NO,                     KC_NO,  KC_NO,  KC_NO,    KC_NO,    KC_NO,    KC_TRNS,
  KC_NO,   KC_NO,   KC_WH_R, KC_BTN3,  KC_WH_L,  KC_NO,    KC_NO, KC_NO,    KC_NO,  KC_NO,  KC_NO,    KC_NO,    KC_NO,    KC_NO,
                                  KC_NO, KC_NO, KC_NO, KC_BTN1,    KC_NO, KC_TRNS, KC_NO, KC_NO
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/lily58/keymaps/datadavd/rules.mk [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+RGBLIGHT_ENABLE = no       # Enable WS2812 RGB underlight.
+SWAP_HANDS_ENABLE = no      # Enable one-hand typing
+OLED_DRIVER_ENABLE= yes     # OLED display
+
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no    # Breathing sleep LED during USB suspend
+
+# If you want to change the display of OLED, you need to change here
+SRC +=  ./lib/rgb_state_reader.c \
+        ./lib/layer_state_reader.c \
```
Comment (drashna) [inline keyboards/lily58/keymaps/datadavd/rules.mk:25]: Sholdn't need this line
```suggestion
```

File: keyboards/lily58/keymaps/datadavd/rules.mk (hunk 3 above)
Comment (DataDavD) [reply keyboards/lily58/keymaps/datadavd/rules.mk]: done

File: keyboards/lily58/keymaps/datadavd/rules.mk (hunk 8 above)
Comment (DataDavD) [reply keyboards/lily58/keymaps/datadavd/rules.mk]: done

**Review by fauxpark — APPROVED** [review]

**Comment by DataDavD** [comment]
Hey friends, I was just wondering if there is anything else I need to do or add. I see Ryan approved the changes, but just wanted to make sure there wasn't anything else I should look forward to changing/adding. 

Thank you for your help and patience with me doing my first PR for QMK. I really appreciate it.

Best,
dd

**Review by drashna — APPROVED** [review]
